### PR TITLE
Added symmetric KL exploration loss

### PR DESCRIPTION
**Pull request by tushartk**

### CODE REVIEW

File: algorithms/utils/action_distributions.py [hunk 1]
```diff
@@ -98,6 +109,9 @@ def _kl_inverse(self, other_log_probs):
     def _kl_symmetric(self, other_log_probs):
         return 0.5 * (self._kl(other_log_probs) + self._kl_inverse(other_log_probs))
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:110]: Can you just rewrite this for the case of a uniform prior?

Something like:
`uniform_prob = 1 / n`
`log_uniform_prob = log(1/n)`
`0.5 * (probs * (log_probs - log_uniform_prob) + uniform_prob * (log_uniform_prob - log_probs))`

Check if this actually works, should be better than the version based on tensor allocation.

Maybe write a test that confirms that this has the same output as the existing version.

File: algorithms/utils/action_distributions.py [hunk 2]
```diff
@@ -98,6 +109,9 @@ def _kl_inverse(self, other_log_probs):
...
 
+    def kl_prior(self):
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:112]: For the uniform prior, rename this to symmetric_kl_with_uniform_prior

File: algorithms/appo/appo.py [hunk 3]
```diff
@@ -118,6 +118,16 @@ def add_cli_args(cls, parser):
         # components of the loss function
         p.add_argument('--entropy_loss_coeff', default=0.003, type=float, help='Coefficient for the exploration component of the loss function.')
         p.add_argument('--value_loss_coeff', default=0.5, type=float, help='Coefficient for the critic loss')
+        p.add_argument('--exploration_loss_type', default='entropy', type=str,
+                       help='Choosing the type of exploration loss', choices=['entropy', 'symmetric_kl'])
+        p.add_argument(
```
[the remark this answers is not in the document]
Comment (tushartk) [reply algorithms/appo/appo.py:123]: Done

File: algorithms/appo/learner.py [hunk 4]
```diff
@@ -399,6 +399,26 @@ def _value_loss(self, new_values, old_values, target, clip_value):
 
         return value_loss
 
+    def maximize_entropy(self, action_distribution):
```
[the remark this answers is not in the document]
Comment (tushartk) [reply algorithms/appo/learner.py:402]: Done

File: algorithms/appo/learner.py [hunk 5]
```diff
@@ -399,6 +399,26 @@ def _value_loss(self, new_values, old_values, target, clip_value):
...
+        entropy = action_distribution.entropy()
+        if self.cfg.entropy_loss_coeff > 0.0:
+            entropy_loss = -self.cfg.entropy_loss_coeff * entropy.mean()
+        else:
+            entropy_loss = 0.0
+        return entropy_loss
+
+    def symmetric_kl_with_uniform_prior(self, action_distribution, mb):
+        if is_continuous_action_space(self.actor_critic.action_space):
```
[the remark this answers is not in the document]
Comment (tushartk) [reply algorithms/appo/learner.py:411]: Done

File: algorithms/appo/learner.py [hunk 6]
```diff
@@ -561,13 +581,13 @@ def _train(self, gpu_buffer, batch_size, experience_size, timing):
                 with timing.add_time('losses'):
                     policy_loss = self._policy_loss(ratio, adv, clip_ratio_low, clip_ratio_high)
 
-                    entropy = action_distribution.entropy()
-                    if self.cfg.entropy_loss_coeff > 0.0:
-                        entropy_loss = -self.cfg.entropy_loss_coeff * entropy.mean()
+                    if self.cfg.exploration_loss_type == 'entropy':
```
[the remark this answers is not in the document]
Comment (tushartk) [reply algorithms/appo/learner.py:584]: Done

File: algorithms/utils/action_distributions.py (hunk 1 above)
Comment (tushartk) [reply algorithms/utils/action_distributions.py:104]: Confirmed the values to be the same. It only was missing `.sum(dim=-1)` which I added.

File: algorithms/utils/action_distributions.py [hunk 7]
```diff
@@ -76,13 +76,24 @@ class CategoricalActionDistribution(Categorical):
...
+    def __init__(self, raw_logits, prior_probs=None):
         """
         Ctor.
         :param raw_logits: unprocessed logits, typically an output of a fully-connected layer
         """
         super().__init__(logits=raw_logits)
 
+        num_categories = raw_logits.shape[-1]
+
+        if prior_probs is None:
```
[the remark this answers is not in the document]
Comment (tushartk) [reply algorithms/utils/action_distributions.py]: Yup, have replaced by the manual code

File: algorithms/appo/appo.py [hunk 8]
```diff
@@ -116,8 +116,11 @@ def add_cli_args(cls, parser):
         p.add_argument('--max_grad_norm', default=4.0, type=float, help='Max L2 norm of the gradient vector')
 
         # components of the loss function
-        p.add_argument('--entropy_loss_coeff', default=0.003, type=float, help='Coefficient for the exploration component of the loss function.')
+        p.add_argument('--exploration_loss_coeff', default=0.003, type=float,
+                       help='Coefficient for the exploration component of the loss function.')
         p.add_argument('--value_loss_coeff', default=0.5, type=float, help='Coefficient for the critic loss')
+        p.add_argument('--exploration_loss', default='entropy', type=str,
+                       help='Choosing the type of exploration loss', choices=['entropy', 'symmetric_kl'])
```
Comment (alex-petrenko) [inline algorithms/appo/appo.py:123]: I think the longer comment is desirable here. You can use my explanation in the issue or in our Slack conversations :)

File: algorithms/appo/learner.py [hunk 9]
```diff
@@ -109,6 +109,16 @@ def __init__(
 
         self.process = Process(target=self._run, daemon=True)
 
+        if is_continuous_action_space(self.action_space):
+            raise NotImplementedError(f'Continuous Action spaces not supported!')
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:113]: Why is this not supported? I think this should only throw when we use KL loss?

Comment (alex-petrenko) [reply algorithms/appo/learner.py:113]: And the exception message should be something like:
"KL-divergence exploration loss is not supported with continuous action spaces. Use entropy exploration loss"

File: algorithms/appo/learner.py [hunk 10]
```diff
@@ -109,6 +109,16 @@ def __init__(
...
+
+        if self.cfg.exploration_loss == 'entropy':
+            self.exploration_loss_func = self.entropy_exploration_loss
+        elif self.cfg.exploration_loss == 'symmetric_kl':
+            self.exploration_loss_func = self.symmetric_kl_exploration_loss
+        else:
+            raise NotImplementedError(f'{self.cfg.exploration_loss} not supported!')
+
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:121]: This is good

Comment (alex-petrenko) [reply algorithms/appo/learner.py:121]: We can actually also avoid checking for `self.cfg.exploration_loss_coeff` later and just check it here.
Just add another if in the beginning:

```
if self.cfg.exploration_loss_coeff == 0.0:
    self.exploration_loss_func = lambda action_distr: 0.0

```

File: algorithms/appo/learner.py [hunk 11]
```diff
@@ -399,6 +409,26 @@ def _value_loss(self, new_values, old_values, target, clip_value):
...
+            entropy_loss = -self.cfg.exploration_loss_coeff * entropy.mean()
+        else:
+            entropy_loss = 0.0
+        return entropy_loss
+
+    def symmetric_kl_exploration_loss(self, action_distribution):
+        if self.cfg.exploration_loss_coeff > 0.0:
+            kl_prior = action_distribution.symmetric_kl_with_uniform_prior()
+            if not torch.isfinite(kl_prior).all():
+                kl_prior = torch.zeros(kl_prior.shape)
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:424]: add log.error here, this should be as visible as possible

Comment (alex-petrenko) [reply algorithms/appo/learner.py:424]: actually you can check this after you do `.mean()`, it should be much faster. If one element is infinite then mean will be infinite as well. Also check for NaN

File: algorithms/appo/learner.py [hunk 12]
```diff
@@ -561,13 +591,8 @@ def _train(self, gpu_buffer, batch_size, experience_size, timing):
...
 
-                    entropy = action_distribution.entropy()
-                    if self.cfg.entropy_loss_coeff > 0.0:
-                        entropy_loss = -self.cfg.entropy_loss_coeff * entropy.mean()
-                    else:
-                        entropy_loss = 0.0
-
-                    actor_loss = policy_loss + entropy_loss
+                    exploration_loss = self.exploration_loss_func(action_distribution)
+                    actor_loss = policy_loss + exploration_loss
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:595]: Good, I like this, looks very clean now

File: algorithms/utils/action_distributions.py [hunk 13]
```diff
@@ -83,6 +84,10 @@ def __init__(self, raw_logits):
         """
         super().__init__(logits=raw_logits)
 
+        num_categories = raw_logits.shape[-1]
+        self.uniform_prob = 1 / num_categories
+        self.log_uniform_prob = math.log(1 / num_categories)
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:89]: you can just `math.log(self.uniform_prob)`

File: algorithms/utils/action_distributions.py [hunk 14]
```diff
@@ -98,6 +103,11 @@ def _kl_inverse(self, other_log_probs):
     def _kl_symmetric(self, other_log_probs):
         return 0.5 * (self._kl(other_log_probs) + self._kl_inverse(other_log_probs))
 
+    def symmetric_kl_with_uniform_prior(self):
+        probs, log_probs = self.probs, self.logits
+        return 0.5 * ((probs * (log_probs - self.log_uniform_prob)).sum(dim=-1)
+                      + (self.uniform_prob * (self.log_uniform_prob - log_probs)).sum(dim=-1))
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:109]: I think this is good. Should work.

Comment (alex-petrenko) [reply algorithms/utils/action_distributions.py:109]: Have you added the same function for Tuple action distributions?

File: algorithms/appo/learner.py (hunk 9 above)
Comment (tushartk) [reply algorithms/appo/learner.py:113]: Oh damn, my mistake, I'll fix it

File: algorithms/appo/learner.py (hunk 11 above)
Comment (tushartk) [reply algorithms/appo/learner.py:424]: the `isfinite()` function checks for both infinity and nan

File: algorithms/utils/action_distributions.py (hunk 14 above)
Comment (tushartk) [reply algorithms/utils/action_distributions.py:109]: Haven't added this, shall add it

**Comment by tushartk** [comment]
Have fixed all issues for now. Will coordinate with Zhehui and start running experiments.

File: algorithms/appo/learner.py [hunk 15]
```diff
@@ -109,6 +109,19 @@ def __init__(
...
+
+        if self.cfg.exploration_loss_coeff == 0.0:
+            self.exploration_loss_func = lambda action_distr: 0.0
+        elif self.cfg.exploration_loss == 'entropy':
+            self.exploration_loss_func = self.entropy_exploration_loss
+        elif self.cfg.exploration_loss == 'symmetric_kl':
+            self.exploration_loss_func = self.symmetric_kl_exploration_loss
+        else:
+            raise NotImplementedError(f'{self.cfg.exploration_loss} not supported!')
+
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:124]: Good!

File: algorithms/appo/learner.py [hunk 16]
```diff
@@ -109,6 +109,19 @@ def __init__(
 
         self.process = Process(target=self._run, daemon=True)
 
+        if is_continuous_action_space(self.action_space) and self.cfg.exploration_loss == 'symmetric_kl':
+            raise NotImplementedError(f'KL-divergence exploration loss is not supported with '
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:113]: String literals start with an f'' if you're not formatting anything :) Just a minor thing

File: algorithms/appo/learner.py [hunk 17]
```diff
@@ -399,6 +412,26 @@ def _value_loss(self, new_values, old_values, target, clip_value):
 
         return value_loss
 
+    def entropy_exploration_loss(self, action_distribution):
+        entropy = action_distribution.entropy()
+        if self.cfg.exploration_loss_coeff > 0.0:
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:417]: These ifs are a bit redundant now, because you already check exploration_loss_coeff in the beginning

File: algorithms/appo/learner.py [hunk 18]
```diff
@@ -399,6 +412,26 @@ def _value_loss(self, new_values, old_values, target, clip_value):
...
+            entropy_loss = -self.cfg.exploration_loss_coeff * entropy.mean()
+        else:
+            entropy_loss = 0.0
+        return entropy_loss
+
+    def symmetric_kl_exploration_loss(self, action_distribution):
+        if self.cfg.exploration_loss_coeff > 0.0:
```
Comment (alex-petrenko) [inline algorithms/appo/learner.py:424]: Same here

File: algorithms/utils/action_distributions.py [hunk 19]
```diff
@@ -193,6 +203,10 @@ def kl_divergence(self, other):
         kl = kls.sum(dim=1)
         return kl
 
+    def symmetric_kl_with_uniform_prior(self):
+        kls = [d.symmetric_kl_with_uniform_prior() for d in self.distributions]
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:207]: Have you tested this? I think this is expected to return a PyTorch scalar, not a list of scalars.
See function kl_divergence()

File: algorithms/appo/learner.py (hunk 16 above)
Comment (tushartk) [reply algorithms/appo/learner.py:113]: Done

File: algorithms/appo/learner.py (hunk 17 above)
Comment (tushartk) [reply algorithms/appo/learner.py:417]: Done

File: algorithms/appo/learner.py (hunk 18 above)
Comment (tushartk) [reply algorithms/appo/learner.py:424]: Done

File: algorithms/utils/action_distributions.py (hunk 19 above)
Comment (tushartk) [reply algorithms/utils/action_distributions.py:207]: If I add  `kl = action_distribution.kl_divergence(action_distribution)` in `test_tuple_distribution` of `test_action_distributions.py` it returns a tensor of size 42 (obviously it'll be all zeros since it is the KL divergence between the same distributions, just added it for the test). Is that correct?

File: algorithms/utils/action_distributions.py [hunk 20]
```diff
@@ -193,6 +203,12 @@ def kl_divergence(self, other):
         kl = kls.sum(dim=1)
         return kl
 
+    def symmetric_kl_with_uniform_prior(self):
+        sym_kls = [d.symmetric_kl_with_uniform_prior().unsqueeze(dim=1) for d in self.distributions]
+        sym_kls = torch.cat(sym_kls, dim=1)
+        sym_kl = sym_kls.sum(dim=1)
+        return sym_kl
+
```
Comment (alex-petrenko) [inline algorithms/utils/action_distributions.py:211]: Yep, this looks good
